### DISCUSSION TITLE
post_233639

### DISCUSSION THREAD

**Discussion by cincuranet** [cincuranet/blog-comments#1261] [Comments]
imported 

https://www.tabsoverspaces.com/id/233639

**Comment by cincuranet** [comment]
Imported comment written by _Максим_ on _2018-06-07T23:19:29_:

<p>Cool</p>

**Comment by Neme12** [comment]
Benchmarks show that `ReadAsync` is more than 2x slower when the async flag is used on FileStream compared to when it's not. Why is that the case if it's supposed to be more efficient?

```c#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.IO;
using System.Threading.Tasks;

BenchmarkRunner.Run<FileStreamBenchmark>();

public class FileStreamBenchmark
{
    private FileStream _fileStream;
    private byte[] _buffer = new byte[1024];

    [Params(false, true)]
    public bool IsAsync { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        byte[] data = new byte[10_000_000];
        new Random(42).NextBytes(data);

        string path = Path.GetTempFileName();
        File.WriteAllBytes(path, data);

        _fileStream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, IsAsync);
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _fileStream.Dispose();
        File.Delete(_fileStream.Name);
    }

    [Benchmark]
    public async Task ReadAsync()
    {
        _fileStream.Position = 0;
        while (await _fileStream.ReadAsync(_buffer) != 0) ;
    }
}
```

Results in .NET 6:
```
|    Method | IsAsync |     Mean |    Error |   StdDev |   Median |
|---------- |-------- |---------:|---------:|---------:|---------:|
| ReadAsync |   False | 17.30 ms | 0.343 ms | 0.938 ms | 17.59 ms |
| ReadAsync |    True | 46.06 ms | 0.903 ms | 1.533 ms | 45.83 ms |
```

I saw a benchmark like this for FileStream in this blogpost: https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#io and wanted to test this myself.

**Comment by cincuranet** [comment]
That's kind of expected. The underlying hardware is the same, and doing it one way or the other shouldn't make one, i.e. the async one, faster, right? However, the _await_ has also a non-zero overhead, but on the other hand gives better efficiency for scalability (speed is not the only efficiency factor). Sadly, here we're doing nothing to take advantage of that better scalability.

To at least improve the numbers a bit we can make the reading buffer bigger, i.e. `private byte[] _buffer = new byte[1024 * 1024];`.

|    Method | IsAsync |     Mean |    Error |  StdDev |
|---------- |-------- |---------:|---------:|--------:|
| ReadAsync |   False | 735.5 us | 10.08 us | 9.43 us |
| ReadAsync |    True | 765.9 us |  6.37 us | 5.96 us |

Still not the same numbers, but closer.
